fdbserver/Coordination.actor.cpp

```diff
@@ -442,32 +448,54 @@ struct LeaderRegisterCollection {
 		if( !self->pStore->exists() )
 			return Void();
 		OnDemandStore &store = *self->pStore;
-		Standalone<RangeResultRef> forwardingInfo = wait( store->readRange( fwdKeys ) );
+		state Future<Standalone<RangeResultRef>> forwardingInfoF = store->readRange(fwdKeys);
+		state Future<Standalone<RangeResultRef>> forwardingTimeF = store->readRange(fwdTimeKeys);
+		wait(success(forwardingInfoF) && success(forwardingTimeF));
+		Standalone<RangeResultRef> forwardingInfo = forwardingInfoF.get();
+		Standalone<RangeResultRef> forwardingTime = forwardingTimeF.get();
 		for( int i = 0; i < forwardingInfo.size(); i++ ) {
 			LeaderInfo forwardInfo;
 			forwardInfo.forward = true;
 			forwardInfo.serializedInfo = forwardingInfo[i].value;
 			self->forward[ forwardingInfo[i].key.removePrefix( fwdKeys.begin ) ] = forwardInfo;
 		}
+		for (int i = 0; i < forwardingTime.size(); i++) {
+			double time = BinaryReader::fromStringRef<double>(forwardingInfo[i].value, IncludeVersion());
+			self->forwardStartTime[forwardingInfo[i].key.removePrefix(fwdTimeKeys.begin)] = time;
+		}
 		return Void();
 	}
 
 	Future<Void> onError() { return actors.getResult(); }
 
 	Optional<LeaderInfo> getForward(KeyRef key) {
 		auto i = forward.find( key );
+		auto t = forwardStartTime.find(key);
 		if (i == forward.end())
 			return Optional<LeaderInfo>();
+		if (t != forwardStartTime.end()) {
+			double forwardTime = t->value;
+			if (SERVER_KNOBS->FORWARD_REQUEST_TOO_OLD > 0 &&
+			    now() - forwardTime > SERVER_KNOBS->FORWARD_REQUEST_TOO_OLD) {
+				TraceEvent(SevWarnAlways, "AccessOldForward")
+				    .detail("ForwardSetSecondsAgo", now() - forwardTime)
+				    .detail("ForwardClusterKey", key);
+			}
+		}
 		return i->value;
 	}
 
 	ACTOR static Future<Void> setForward(LeaderRegisterCollection *self, KeyRef key, ClusterConnectionString conn) {
+		double forwardTime = now();
```

Comment:
Now look at this, I'd better use read version instead of a local time stamp since the clocks are not sync'd.

Reply:
But then that will make the expiry check in `getForward` became async. Or I can have a `AsyncVar` which is updated periodically which can be used to do the check.

Reply:
I think using the version is a bad idea since it breaks the abstraction boundary. Since the expiry check doesn't need to be super accurate, I think using local clock should be fine.